07-errors.md

```diff
@@ -0,0 +1,301 @@
+   `ExitCode::SysErrSenderInvalid` abnormal exit code.
+2. If an actor calls `vm::abort` with a system abnormal exit code, the send to that actor
+   will fail with the `ExitCode::ErrIllegalActor` _syscall error_. If this is a top-level send (the
+   execution of an on-chain message), the `ExitCode::SysErrIllegalActor` system exit code will be
+   recorded as the messages exit code.
+3. If a message runs out of gas, the VM will return a receipt with the `ExitCode::SysErrOutOfGas`
+   abnormal exit code.
+
+Changes from pre-FVM Filecoin:
+
+1. Removed `ExitCode::SysErrInvalidMethod` (3). Dispatch is now handled inside actor code.
+2. Removed `ExitCode::SysErrForbidden` (8). Caller verification is a userspace operation, we'll now
+   exit with `ExitCode::ErrForbidden`.
+4. Replaced `ExitCode::SysErrIllegalArgument` with `ExitCode::SysErrAssertionFailed` to more
+   accurately reflect its usage. When this appears on-chain, it means that a message hit some
+   internal Filecoin assert that shouldn't happen.
+
+### Standard "Abnormal Exit" Codes
+
+Standard abnormal exit codes are codes returned by actors to indicate common abnormal exit conditions.
+
+| Code  | Status                 | Description                                         |
+|-------|------------------------|-----------------------------------------------------|
+| 16    | `ErrIllegalArgument`   | invalid message parameters                          |
+| 17    | `ErrNotFound`          | message referenced something that doesn't exist     |
+| 18    | `ErrForbidden`         | operation forbidden                                 |
+| 19    | `ErrInsufficientFunds` | insufficient funds for operation                    |
+| 20    | `ErrIllegalState`      | the actor is in an illegal state                    |
+| 21    | `ErrSerialization`     | actor failed to serialize or deserialize some state |
+| 22    | `ErrInvalidMethod`     | specified method does not exist                     |
+| 23    | `ErrUnspecified`       | the actor failed with an unspecified error          |
+| 24-31 | reserved               |                                                     |
+
+Changes from pre-FVM Filecoin:
+
+1. Added `ExitCode::ErrInvalidMethod` (replacing `ExitCode::SysErrInvalidMethod`).
+2. Added `ExitCode::ErrUnspecified`. This generally means "there's a bug and I have no idea what
+   went wrong". The alternative would be for the actor to _panic_ (leading to an
+   `ExitCode::ErrIllegalActor` syscall error), but it may be useful to distinguish between
+   "crashing" and "something unknown went wrong".
+
+## Syscall Error Numbers
+
+Syscall error numbers (`ErrorNumber`) are returned by syscalls to actors. They indicate that the
+syscall failed (without any side effects). The actor may choose to handle the error and continue, or
+abort (usually with a standard "abnormal exit" code).
+
+A return value of "0" means that the syscall succeeded.
```

Comment:
This is a bit confusing; yes, it _implies_ that, but it actually _means_ that no return data is being returned.

Reply:
No? Zero here just means that the syscall succeeded and that the caller should check the "out" parameters (if any) for a return value.

Reply:
Duh, I was thinking about the actor's return, not the syscall return. Please disregard my comment.